odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2372 @@
+            self.data.data.copy_to_host(
+                out.ctypes.data_as(np.ctypeslib.ctypes.c_void_p),
+                self.size * self.itemsize)
+            return out
+
+    @property
+    def data_ptr(self):
+        """A raw pointer to the data container.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.one()
+        >>> x.data_ptr  # doctest: +SKIP
+        47259975936
+        """
+        return self.data.ptr
+
+    def __eq__(self, other):
+        """Return ``self == other``.
+
+        Parameters
+        ----------
+        other :
+            Object to be compared with ``self``.
+
+        Returns
+        -------
+        equals : bool
+            ``True`` if all entries of ``other`` are equal to this
+            tensor's entries, ``False`` otherwise.
+
+        Notes
+        -----
+        The element-by-element comparison is performed on the CPU,
+        i.e. it involves data transfer to host memory, which is slow.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> same_x = r3.element([1, 2, 3])
+        >>> y = r3.element([-1, -2, -3])
+        >>> x == same_x
+        True
+        >>> x == y
+        False
+
+        Space membership matters:
+
+        >>> int_spc = odl.tensor_space(3, impl='cupy', dtype=int)
+        >>> x_int = int_spc.element([1, 2, 3])
+        >>> x == x_int
+        False
+        """
+        if other is self:
+            return True
+        elif other not in self.space:
+            return False
+        else:
+            return bool((self.data == other.data).all())
+
+    def copy(self):
+        """Create an identical (deep) copy of this tensor.
+
+        Returns
+        -------
+        copy : `CupyTensor`
+            A deep copy.
+
+        Examples
+        --------
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> y = x.copy()
+        >>> y
+        rn(3, impl='cupy').element([ 1.,  2.,  3.])
+        >>> x == y
+        True
+        >>> x is y
+        False
+        """
+        return self.space.element(self.data.copy())
+
+    def __getitem__(self, indices):
+        """Access values of this tensor.
+
+        Parameters
+        ----------
+        indices : index expression
+            The position(s) that should be accessed.
+
+        Returns
+        -------
+        values : scalar or `cupy.core.core.ndarray`
+            The value(s) at the index (indices).
+
+        Examples
+        --------
+        Indexing rules follow roughly the Numpy style, as far (or "fancy")
+        as supported:
+
+        >>> r5 = odl.rn(5, impl='cupy')
+        >>> x = r5.element([1, 2, 3, 4, 5])
+        >>> x[1:4]
+        rn(3, impl='cupy').element([ 2.,  3.,  4.])
+        >>> x[::2]
+        rn(3, impl='cupy').element([ 1.,  3.,  5.])
+
+        The returned views are writable, so modificatons alter the
```

Comment:
Use "view"

Reply:
Also mention that this is not always the case (i guess?)

Reply:
Not for index arrays and boolean arrays, right.